packages/core/src/storage/namespace.ts

```diff
@@ -0,0 +1,396 @@
+      }
+    }
+  }
+  return undefined;
+}
+
+/**
+ * From the given layout types, gets the type matching the given type label.
+ */
+function findTypeWithLabel(types: Record<string, TypeItem>, label: string) {
+  for (const type of Object.values(types)) {
+    if (type.label === label) {
+      return type;
+    }
+  }
+  return undefined;
+}
+
+/**
+ * Makes a modified copy of the solc input to add state variables in each contract for namespaced struct definitions,
+ * so that the compiler will generate their types in the storage layout.
+ *
+ * This deletes all functions for efficiency, since they are not needed for storage layout.
+ * We also need to delete modifiers and immutable variables to avoid compilation errors due to deleted
+ * functions and constructors.
+ *
+ * Also sets the outputSelection to only include storageLayout and ast, since the other outputs are not needed.
+ *
+ * @param input The original solc input.
+ * @param output The original solc output.
+ * @returns The modified solc input with storage layout that includes namespaced type information.
+ */
+export function makeNamespacedInputCopy(input: SolcInput, output: SolcOutput): SolcInput {
+  const modifiedInput: SolcInput = JSON.parse(JSON.stringify(input));
+
+  modifiedInput.settings = {
+    outputSelection: {
+      '*': {
+        '*': ['storageLayout'],
+        '': ['ast'],
+      },
+    },
+  };
+
+  for (const [sourcePath] of Object.entries(modifiedInput.sources)) {
+    const source = modifiedInput.sources[sourcePath];
+    if (source.content === undefined) {
+      continue;
+    }
+
+    // Collect all contract definitions
+    const contractDefs = [];
+    for (const contractDef of findAll('ContractDefinition', output.sources[sourcePath].ast)) {
+      contractDefs.push(contractDef);
+    }
+
+    // Iterate backwards so we can delete source code without affecting remaining indices
+    for (let i = contractDefs.length - 1; i >= 0; i--) {
+      const contractDef = contractDefs[i];
+      const nodes = contractDef.nodes;
+      for (let j = nodes.length - 1; j >= 0; j--) {
+        const node = nodes[j];
+        if (
+          isNodeType('FunctionDefinition', node) ||
+          isNodeType('ModifierDefinition', node) ||
+          (isNodeType('VariableDeclaration', node) && node.mutability === 'immutable')
+        ) {
+          const orig = Buffer.from(source.content);
+
+          const [start, length] = node.src.split(':').map(Number);
+          let end = start + length;
+
+          // If the next character is a semicolon (e.g. for immutable variables), delete it too
+          if (end + 1 < orig.length && orig.subarray(end, end + 1).toString() === ';') {
+            end += 1;
+          }
+
+          // Delete the source code segment
+          const buf = Buffer.concat([orig.subarray(0, start), orig.subarray(end)]);
+
+          source.content = buf.toString();
+        } else if (isNodeType('StructDefinition', node)) {
```

Comment:
I see that this part is going to break if we do what I said above about filling with spaces. What we could do is to build a contract like `contract ___ is ${contractDef.name} { ${structName} ${variableName}; }` and append it at the end of the file outside the for loop...

Open to other ideas. To be honest, we should optimize prematurely, but avoiding copying the file contents over and over again seems like a no brainer.

Reply:
Ok so this idea of appending a new contract at the end of the file is probably not going to work because our goal is to add the namespace struct into the types of the `storageLayout` output for the original contract. So we have to somehow add the variable inside the original contract.

An alternative to copying the contents over and over again is to only do it at the end when we convert the source back to a string. We can construct an array of... something and then at the end concatenate it all into a string.